using_restricted_array.rb

```diff
@@ -7,43 +7,96 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  while array[i] != nil
+    i += 1
+  end
+  return i
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  while array[i] != nil
+    print i
+    i += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  while i < length
+    if array[i] == value_to_find
+      return true
+    end
+    i += 1
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  largest_value = array[0]
+  while i < length
+    if array[i] > largest_value
+      largest_value = array[i]
+    end
+    i += 1
+  end
+  return largest_value
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  smallest_value = array[0]
+  while i < length
+    if array[i] < smallest_value
+      smallest_value = array[i]
+    end
+    i += 1
+  end
+  return smallest_value
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  j = length - 1
+  while i < j
+    last = array[j]
+    first = array[i]
+    array[i] = last
+    array[j] = first
+    i += 1
+    j -= 1
+  end
+  return array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def binary_search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
-end
+  low = 0
+  high = length - 1
 
+  while low < high
```

Comment:
Please take a look at the updated binary_search slide. 
Here's the considerations I was missing in the algorithm on the slide before the update: (1) what if there is only one element in the array and it is the value_to_find? In this case, it used to return false. (To fix this, add an additional if statement after the loop ends, to check if array[low] is the value_to_find.) (2) Consider the scenario where there are two elements in the array and the value_to_find is the second element in the array. In the first iteration, low = 0, high = 1, mid=0. Value is not at index 0. So, low = mid (still 0). On re-entering the loop, mid would be still 0. So we had an infinite loop. (To fix this, update high to be mid -1, if value_to_find is less than mid. And update low = mid+1, if value_to_find is greater than mid.)